vendor/blockchain_contracts/src/ethereum/rfc003/erc20_htlc.rs

```diff
@@ -0,0 +1,219 @@
+use regex::Regex;
+
+#[derive(Debug, Clone)]
+pub struct Erc20Htlc {
+    refund_timestamp: Timestamp,
+    refund_address: Address,
+    redeem_address: Address,
+    secret_hash: SecretHash,
+    token_contract_address: Address,
+    amount: U256,
+}
+
+impl Erc20Htlc {
+    const CONTRACT_CODE_TEMPLATE: &'static str =
+        include_str!("./templates/out/erc20_contract.asm.hex");
+    const SECRET_HASH_PLACEHOLDER: &'static str =
+        "1000000000000000000000000000000000000000000000000000000000000001";
+    const REFUND_TIMESTAMP_PLACEHOLDER: &'static str = "20000002";
+    const REDEEM_ADDRESS_PLACEHOLDER: &'static str = "3000000000000000000000000000000000000003";
+    const REFUND_ADDRESS_PLACEHOLDER: &'static str = "4000000000000000000000000000000000000004";
+    const AMOUNT_PLACEHOLDER: &'static str =
+        "5000000000000000000000000000000000000000000000000000000000000005";
+    const TOKEN_CONTRACT_ADDRESS_PLACEHOLDER: &'static str =
+        "6000000000000000000000000000000000000006";
+
+    const DEPLOY_HEADER_TEMPLATE: &'static str =
+        include_str!("./templates/out/erc20_deploy_header.asm.hex");
+    const CONTRACT_START_POSITION_PLACEHOLDER: &'static str = "1001";
+    const CONTRACT_LENGTH_PLACEHOLDER: &'static str = "2002";
+
+    pub fn new(
+        refund_timestamp: Timestamp,
+        refund_address: Address,
+        redeem_address: Address,
+        secret_hash: SecretHash,
+        token_contract_address: Address,
+        amount: U256,
+    ) -> Self {
+        Self {
+            refund_timestamp,
+            refund_address,
+            redeem_address,
+            secret_hash,
+            token_contract_address,
+            amount,
+        }
+    }
+
+    /// Constructs the payload for funding an `Erc20` HTLC located at the given
+    /// address.
+    pub fn funding_tx_payload(&self, htlc_contract_address: Address) -> Bytes {
+        let transfer_fn_abi = base16!("A9059CBB");
+        let htlc_contract_address = <[u8; 20]>::from(htlc_contract_address);
+        let amount = <[u8; 32]>::from(self.amount);
+
+        let mut data = [0u8; 4 + 32 + 32];
+        data[..4].copy_from_slice(transfer_fn_abi);
+        data[16..36].copy_from_slice(&htlc_contract_address);
+        data[36..68].copy_from_slice(&amount);
+
+        Bytes::from(data.to_vec())
+    }
+
+    pub fn token_contract_address(&self) -> Address {
+        self.token_contract_address
+    }
+
+    pub fn deployment_gas_limit(&self) -> U256 {
+        let bytes: Bytes = self.compile_to_hex().into();
+        let n_bytes = bytes.0.len();
+        let gas_per_byte = 200;
+
+        U256::from(100_000 + n_bytes * gas_per_byte)
+    }
+
+    pub fn tx_gas_limit() -> U256 {
+        U256::from(100_000)
+    }
+
+    pub fn fund_tx_gas_limit() -> U256 {
+        U256::from(100_000)
+    }
+
+    pub fn compile_template_to_hex() -> String {
+        let code_length = Self::CONTRACT_CODE_TEMPLATE.len() / 2; // In hex, each byte is two chars
```

Comment:
🔧 Would be nicer if we just had `Vec<u8>` here so we don't have to divide stuff by 2 everywhere.

Reply:
It does not look like I could do something as easy as I do with regex to search a `Vec<u8>` in a `Vec<u8>` so I prefer to keep `String` and `regex` except if you have a better idea.

Reply:
There's actually a regex::bytes if you were interested: https://docs.rs/regex/1.1.6/regex/bytes/index.html

Reply:
done!